### PR TITLE
test: Use coverage over pytest-cov

### PR DESCRIPTION
**Pull request by matthewfeickert**
# Description

Following Anthony Sottile's video [I don't use pytest-cov (intermediate) anthony explains 489](https://youtu.be/sPgvHGkmd0U), this PR drops `pytest-cov` in favor of just managing coverage ourselves with `coverage[toml]`. This implements running the tests with coverage as and **optional** part of the `tests` `nox` session (enabled with adding `coverage` as a positional argument) and adds generating the coverage report as a separate step with `nox -s coverage`:

* Simplifies the reporting of pytest runs to show the information that is most relevant.
* Speeds up running the pytest tests as coverage takes additional time to properly check all branches. For the full test suite the result is a speed up of about a minute.

`pytest` with `coverage`: `03:54`
```console
$ nox -s tests-3.10 -- coverage
...
================================================================== 1435 passed, 43 skipped, 12 xfailed in 234.80s (0:03:54) ==================================================================

```

vs.

`pytest` without `coverage`: `03:04`
```console
$ nox -s tests-3.10
...
================================================================== 1435 passed, 43 skipped, 12 xfailed in 184.05s (0:03:04) ==================================================================

```

* Removes `.coveragerc` in favor of placing the information in `pyproject.toml` with the `coverage run --omit` option via

```
omit = ["*/pyhf/typing.py"]
```

* Sort the coverage report in ascending order of coverage, making it easier to visually group the files that need improved coverage.

* Add coverage report viewing steps to the CI.

# Checklist Before Requesting Reviewer

- [x] Tests are passing
- [x] "WIP" removed from the title of the pull request
- [x] Selected an Assignee for the PR to be responsible for the log summary

# Before Merging

For the PR Assignees:

- [x] Summarize commit messages into a comprehensive review of the PR

```
* Remove 'pytest-cov' in favor of 'coverage[toml]' in 'test' extra.
* Remove .coveragerc in favor of setting coverage options in pyproject.toml.
* Make running without coverage the default for the tests nox session. Running under coverage
  can be enabled with an optional 'coverage' positional argument.
* Add coverage session to nox that displays a coverage report sorted by ascending coverage
  and generates XML and HTML versions of the report.
* Update CI workflows to run with coverage only if needed. 
   - Apply stylistic changes to the order of arguments to pytest to match across the codebase.
* Add steps to CI to generate the stdout coverage report and XML report after each segment
  of the tests.
* Add section to development docs on coverage.
```

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -47,14 +47,19 @@ jobs:
     - name: List installed Python packages
       run: python -m pip list
 
-    - name: Test with pytest
+    - name: Test with pytest and coverage
       run: |
-        pytest --ignore tests/benchmarks/ --ignore tests/contrib --ignore tests/test_notebooks.py
+        coverage run --module pytest --ignore tests/contrib --ignore tests/benchmarks --ignore tests/test_notebooks.py
```
Comment (matthewfeickert) [inline .github/workflows/ci.yml:52]: Need to now manually run under `coverage` and the `--module` is just how you tell it to run something like `pytest`, and not a local runner script.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -47,14 +47,19 @@ jobs:
...
 
     - name: Launch a tmate session if tests fail
       if: failure() && github.event_name == 'workflow_dispatch'
       uses: mxschmitt/action-tmate@v3
 
+    - name: Coverage report for core project
+      run: |
+        coverage report
+        coverage xml
```
Comment (matthewfeickert) [inline .github/workflows/ci.yml:61]: Need to now generate the report for uploading, but the report is now a clear stage by itself.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -78,13 +88,19 @@ jobs:
...
     - name: Test docstring examples with doctest
       if: matrix.python-version == '3.10'
-      run: pytest src/ README.rst
+      run: coverage run --data-file=.coverage-doctest --module pytest src/ README.rst
+
+    - name: Coverage report for doctest only
+      if: matrix.python-version == '3.10'
+      run: |
+        coverage report --data-file=.coverage-doctest
+        coverage xml --data-file=.coverage-doctest -o doctest-coverage.xml
```
Comment (matthewfeickert) [inline .github/workflows/ci.yml:97]: Rename the report for doctest as it is being run over the local files, but the tests are covering the installed library. So just highlighting that the reports should be different.

File: setup.py [hunk 4]
```diff
@@ -29,7 +29,7 @@
         + [
             'scikit-hep-testdata>=0.4.11',
             'pytest>=6.0',
-            'pytest-cov>=2.5.1',
+            'coverage[toml]>=6.0.0',
```
Comment (matthewfeickert) [inline setup.py:32]: Need `toml` extra to be able to insure reading of `pyproject.toml`.

**Review by matthewfeickert — COMMENTED** [review]
Just notes to make the review easier hopefully.

**Comment by matthewfeickert** [comment]
Relevant RTD build: https://pyhf--2054.org.readthedocs.build/en/2054/development.html#coverage

File: noxfile.py [hunk 5]
```diff
@@ -58,21 +67,41 @@ def _contrib(session):
         return
 
     if session.posargs:
-        session.run("pytest", *session.posargs)
+        session.run(*runner_commands, *session.posargs)
     else:
         # defaults
+        default_runner_commands = runner_commands.copy()
+        if "--append" in default_runner_commands:
+            default_runner_commands.pop(default_runner_commands.index("--append"))
```
Comment (matthewfeickert) [inline noxfile.py:75]: Perform a copy here so that when `--append` is `pop`ed from the list it doesn't get `pop`ed from the `posargs` list that will be used in `_contrib` and so then won't clobber the previous `.coverage` file from the rest of the tests.

File: .github/workflows/dependencies-head.yml [hunk 6]
```diff
@@ -91,9 +91,9 @@ jobs:
         python -m pip list
     - name: Test with pytest
       run: |
-        pytest --ignore tests/benchmarks/ --ignore tests/contrib --ignore tests/test_notebooks.py
+        pytest --ignore tests/contrib --ignore tests/benchmarks --ignore tests/test_notebooks.py
 
-  uproot4:
+  uproot5:
```
Comment (kratsg) [inline .github/workflows/dependencies-head.yml:96]: should be a separate PR

Comment (matthewfeickert) [reply .github/workflows/dependencies-head.yml:96]: I can do it if you want, but this is a cosmetic change in that https://github.com/scikit-hep/uproot4 resolves to https://github.com/scikit-hep/uproot5.

**Review by kratsg — APPROVED** [review]
